### PR TITLE
Dev aj

### PR DESCRIPTION
**Pull request by ajdema**
Implemented GenerateMoves for A simple Game (jan 4x4). A move is represented as a 32 bit int where bits 7-4 are the start position and bits 3-0 are the ending position.

### CODE REVIEW

File: src/mjan.c [hunk 1]
```diff
@@ -129,6 +129,7 @@ void InitializeGame(void) {
     // then leave gPositionToStringFunPtr as NULL, i.e.,
     // delete this next line and also feel free
     // to delete the PositionToString function.
+    int gCurrPlayer = 1; // Current player
```
Comment (BnjmnZmmrmn) [inline src/mjan.c]: remove, use `generic_hash_turn`

File: src/mjan.c [hunk 2]
```diff
@@ -158,6 +159,33 @@ MOVELIST *GenerateMoves(POSITION position) {
 
         See the function CreateMovelistNode in src/core/misc.c
     */
+    char* board = (char*) SafeMalloc(16 * sizeof(char));
+    char* board = generic_hash_unhash(position, board);
+
+    int player = generic_hash_turn(position);
+    char piece = (player == 1) ? 'w' : 'b';
+    
+    int directions[4] = {-4, 4, -1, 1};
```
Comment (BnjmnZmmrmn) [inline src/mjan.c:168]: leave comment to explain up down left right

File: src/mjan.c [hunk 3]
```diff
@@ -158,6 +159,33 @@ MOVELIST *GenerateMoves(POSITION position) {
...
+        if (board[i] == piece) {
+
+            for (int d = 0; d < 4; d++) {
+                int new_pos = i + directions[d]
+
+                if (new_pos >= 0 && new_pos < 16 && board[new_pos] == '-') {
+                    if (directions[d] == -1 && (i % 4 == 0)) continue;
+                    if (directions[d] == 1 && (i % 4 == 3)) continue;
+
+                    int encoded_move = ((i & 0x0F) << 4) | (new_pos & 0x0F);
```
Comment (BnjmnZmmrmn) [inline src/mjan.c]: what are `&` operations for? `<<` leaves 0s

Comment (BnjmnZmmrmn) [reply src/mjan.c]: ah nvm `&` necessary for `new_pos`

File: src/mjan.c [hunk 4]
```diff
@@ -172,7 +200,21 @@ MOVELIST *GenerateMoves(POSITION position) {
  * the POSITION typedef in src/core/types.h.
  */
 POSITION DoMove(POSITION position, MOVE move) {
-    return 0;
+    char* board = (char*) SafeMalloc(16 * sizeof(char));
+    char* board = generic_hash_unhash(position, board);
+    int curr_player = generic_hash_turn(position);
+
+    int start_pos = ((move >> 4) & 0x0F) + 1;
```
Comment (BnjmnZmmrmn) [inline src/mjan.c]: what are `+1`s for?

File: src/mjan.c [hunk 5]
```diff
@@ -172,7 +200,21 @@ MOVELIST *GenerateMoves(POSITION position) {
...
+    int end_pos = (move & 0x0F) + 1;
+
+    char piece = board[start_pos];
+    board[end_pos] = piece;
+    boad[start_pos] = '-';
```
Comment (BnjmnZmmrmn) [inline src/mjan.c]: `board` me thinks

**Review by BnjmnZmmrmn — CHANGES_REQUESTED** [review]
please see comments

File: src/mjan.c (hunk 3 above)
Comment (ajdema) [reply src/mjan.c]: Valid

File: src/mjan.c (hunk 5 above)
Comment (ajdema) [reply src/mjan.c]: mm good idea

**Review by BnjmnZmmrmn — APPROVED** [review]
lgtm

**Pull request merged**